packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts

```diff
@@ -1,470 +1,630 @@
+                    buyQuoteFunctionSelector: curve.buyQuoteFunctionSelector,
+                },
+                new BigNumber(fromTokenIdx),
+                new BigNumber(toTokenIdx),
+                makerFillAmounts,
+            ],
+        });
+    }
+
+    public getBalancerSellQuotes(
+        poolAddress: string,
         makerToken: string,
         takerToken: string,
         takerFillAmounts: BigNumber[],
-        bancorService: BancorService,
-    ): SourceQuoteOperation<BancorFillData> {
-        return {
-            source: ERC20BridgeSource.Bancor,
-            encodeCall: _contract => {
-                return '0x';
-            },
-            handleCallResultsAsync: async (_contract, _callResults) => {
-                return Promise.all(
-                    takerFillAmounts.map(async amt => bancorService.getQuoteAsync(takerToken, makerToken, amt)),
-                );
-            },
-        };
-    },
-    getBalancerSellQuotes(pool: BalancerPool, takerFillAmounts: BigNumber[]): SourceQuoteOperation<BalancerFillData> {
-        return {
+    ): SourceQuoteOperation<BalancerFillData> {
+        return new SamplerContractOperation({
             source: ERC20BridgeSource.Balancer,
-            ...samplerOperations.constant(
-                takerFillAmounts.map(amount => ({
-                    amount: computeBalancerSellQuote(pool, amount),
-                    fillData: { poolAddress: pool.id },
-                })),
-            ),
-        };
-    },
-    getBalancerBuyQuotes(pool: BalancerPool, makerFillAmounts: BigNumber[]): SourceQuoteOperation<BalancerFillData> {
-        return {
+            fillData: { poolAddress },
+            contract: this._samplerContract,
+            function: this._samplerContract.sampleSellsFromBalancer,
+            params: [poolAddress, takerToken, makerToken, takerFillAmounts],
+        });
+    }
+
+    public getBalancerBuyQuotes(
+        poolAddress: string,
+        makerToken: string,
+        takerToken: string,
+        makerFillAmounts: BigNumber[],
+    ): SourceQuoteOperation<BalancerFillData> {
+        return new SamplerContractOperation({
             source: ERC20BridgeSource.Balancer,
-            ...samplerOperations.constant(
-                makerFillAmounts.map(amount => ({
-                    amount: computeBalancerBuyQuote(pool, amount),
-                    fillData: { poolAddress: pool.id },
-                })),
-            ),
-        };
-    },
-    getMStableSellQuotes(makerToken: string, takerToken: string, takerFillAmounts: BigNumber[]): SourceQuoteOperation {
-        return {
+            fillData: { poolAddress },
+            contract: this._samplerContract,
+            function: this._samplerContract.sampleBuysFromBalancer,
+            params: [poolAddress, takerToken, makerToken, makerFillAmounts],
+        });
+    }
+
+    public async getBalancerSellQuotesAsync(
```

Comment:
suggestion: make this quite clear that it is an off-chain lookup, perhaps by function name.